cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -633,16 +667,73 @@ private module ImplCommon {
         result = "CcCall(" + call + ", " + i + ")"
       )
     }
+
+    override predicate relevantFor(DataFlowCallable callable) {
+      recordDataFlowCallSite(getCall(), callable)
+    }
+
+    DataFlowCall getCall() { this = TSpecificCall(result, _, _) }
   }
 
   class CallContextSomeCall extends CallContextCall, TSomeCall {
     override string toString() { result = "CcSomeCall" }
+
+    override predicate relevantFor(DataFlowCallable callable) {
+      exists(ParameterNode p | this = TSomeCall(p, _) and p.getEnclosingCallable() = callable)
+    }
   }
 
   class CallContextReturn extends CallContext, TReturn {
     override string toString() {
       exists(DataFlowCall call | this = TReturn(_, call) | result = "CcReturn(" + call + ")")
     }
+
+    override predicate relevantFor(DataFlowCallable callable) {
+      exists(DataFlowCall call | this = TReturn(_, call) and call.getEnclosingCallable() = callable)
+    }
+  }
+
+  /**
+   * A call context that is relevant for pruning local flow.
+   */
+  abstract class LocalCallContext extends TLocalFlowCallContext {
+    abstract string toString();
+
+    /** Holds if this call context is relevant for `callable`. */
+    abstract predicate relevantFor(DataFlowCallable callable);
+  }
+
+  class LocalCallContextAny extends LocalCallContext, TAnyLocalCall {
+    override string toString() { result = "LocalCcAny" }
+
+    override predicate relevantFor(DataFlowCallable callable) { any() }
+  }
+
+  class LocalCallContextSpecificCall extends LocalCallContext, TSpecificLocalCall {
+    LocalCallContextSpecificCall() { this = TSpecificLocalCall(call) }
+
+    DataFlowCall call;
+
+    DataFlowCall getCall() { result = call }
+
+    override string toString() { result = "LocalCcCall(" + call + ")" }
+
+    override predicate relevantFor(DataFlowCallable callable) { relevantLocalCCtx(call, callable) }
+  }
+
+  private predicate relevantLocalCCtx(DataFlowCall call, DataFlowCallable callable) {
+    exists(Node n | n.getEnclosingCallable() = callable and isUnreachableInCall(n, call))
+  }
+
+  /**
+   * Gets the local call context given the call context and the callable that
+   * the contexts apply to.
+   */
+  LocalCallContext getLocalCallContext(CallContext ctx, DataFlowCallable callable) {
```

Comment:
Do you think it makes sense to cache this predicate?

Reply:
It might.  I don't think it's crucial, though.